### ISSUE TITLE
Dear users, I need your opinion on something

### ISSUE THREAD

**Issue by Minoru** [akrennmair/newsbeuter#249]
While working on the fix for #193 I've run into a problem of design that I want to discuss with you all.

The gist of #193 is as follows: when you have two feeds that partially consist of identical items (think of feeds for "Asian news" and "Chinese news", for example), Newsbeuter only displays these items in one of the feeds and not the other.

That's easy enough to fix by itself, but it begets problems with reading, flagging and deleting. If an item belongs to two feeds, how should Newsbeuter behave when I delete the item in one of the feeds? Should this behaviour differ from the behaviour when the item is being marked as read, or when its flags are changed?

I believe the behaviour should be consistent, i.e. it should be the same for reading, deleting and flagging. But I can't decide if items should be treated as separate or as one. So I'm thinking of adding an option, say, `treat-identical-items-as-one`.

But then we have to decide on default value, which proves to be just as difficult. Here's the problem with each value.

* `yes`. Seems reasonable, right? You don't want to go over all your feeds to delete the same items. You don't want to see duplicates in query feeds. But that setting will obscure the very fact that the items were originally present in a number of feeds (because when you delete it from one feed, it will disappear from all of them). This will be a case of "silent breakage", something I passionately hate and wouldn't do to users if at all possible.
* `no`. If items are treated as separate, you have to deal with each of them individually. No shortcuts to delete from all feeds, nothing like that. On the other hand, you get to flag each item differently depending on the feed, and can mark it as read in one feed and still keep it unread in another.

So which of these values would **you** choose? Or maybe you think that the whole idea of such a setting is silly, and we should do something else? Am I missing some relevant usage scenarios or corner cases? Please say so!

You can comment right here, or you can email me (see my profile for email), or you may contact me on irc.freenode.net (nickname's Minoru, of course; if I'm not there, use MemoServ; oh, and do join #newsbeuter channel to help your fellow users with their questions, even though we don't get much of those.)

Thank you, and good day to you all.

**Comment by kriskropd** [comment]
Mine would be set to 'yes' in my config.

I agree it's a hard default to pick as it is a crucial preference per user, but I would think it's a minority who prefer their newsbeuter treats each duplicate individually. It might actually be requested later that only some feeds treat duplicates as-one while other feeds treat duplicates individually. I could see why someone may want to have that kind of configurable flexibility later.

**Comment by ntnn** [comment]
FTF: Thanks to @kriskropd ping on freenode.

Personally I'd welcome the feature as an option and binding.

I'd enable and disable it as needed (or use as keybind), simply because I have multiple news
stations with multiple feeds (e.g. heise with its subfeeds, where the
full feed amounts to about 30 items per day).
When I go through a subchannel I'd love to have a bind which lets me
mark identical news items mark as read - however when simply mark the
main channel as read I don't want to miss out on the important
subchannels.

That option would be a great help in that use-case.

**Comment by ntnn** [comment]
FTF: Thanks to @kriskropd ping on freenode.

Personally I'd welcome the feature as an option (and keybind).

I'd enable and disable it as needed, simply because I have multiple news stations with multiple feeds (e.g. heise with its subfeeds, where the full feed amounts to about 30 items per day).
When I go through a subchannel I'd love to have a bind which lets me mark identical news items mark as read - however when simply mark the main channel as read I don't want to miss out on the important subchannels.

That option would be a great help in that use-case.

**Comment by ntnn** [comment]
FTF: Thanks to @kriskropd ping on freenode.

Personally I'd welcome the feature as an option (and keybind).

I'd enable and disable it as needed, simply because I have multiple news
stations with multiple feeds (e.g. heise with its subfeeds, where the
full feed amounts to about 30 items per day).
When I go through a subchannel I'd love to have a bind which lets me
mark identical news items mark as read - however when simply mark the
main channel as read I don't want to miss out on the important
subchannels.

That option would be a great help in that use-case.

**Comment by Captain-Rage** [comment]
I agree with the rationale that kriskropd wrote ("least labour"). Personally it would suit me better to have identical entries treated collectively, instead of having to handle them separately, from a pragmatic point of view. That being said, I never subcribed to any feeds that would produce duplicate entities.

**Comment by kriskropd** [comment]
Expandng on the idea of keybinding that @ntnn suggested - It seems like an arbitrary, but necessary discrimination for *some* users. One solution might be to include all duplicate articles in the database and have the feature to hide/mark duplicates built into the UI exclusively. This would exclude the necessity of storing information about duplication within the database entirely.

When users wish to delete a duplicated item, collectively delete the items in the database, even if hidden in the UI. Then if a users prefer for duplicates to be displayed, treat them individually so they only are deleted in that particular feed and not in any other feeds. This way it isn't detrimental to any users to have the default value set to `no` and they can easily switch back and forth while previewing newsbeuter for the first time.

**Comment by Minoru** [comment]
Okay, so two votes for "yes" by default, so far. Cool.

@ntnn, by keybind you mean a toggle that changes the value of the setting? We'll need to display its current state, then. Any ideas about a succinct way to do that?

@kriskropd's extension of this idea sounds really good, but only if we talk about query feeds where duplicate items are readily visible. But if you're viewing, say, "China news", you don't know which items are unique to this feed and which ones are shared with "Asia news". I have no idea what to do in this case.

**Comment by kriskropd** [comment]
In the case that a user wants duplicates displayed AND marked AND treated individually, but isn't using query-feeds - would it be too much of a performance hit to search the DB for matching GUIDs where feedURLs share the same domain address? In my experience, thats the only way that exact GUID duplicates are found, when feeds share the same sources. That said, I can imagine some custom, local script generating duplicates too and this solution wouldn't find those, I think.

**Comment by ntnn** [comment]
@Minoru that or separate commands, i.e. `r` is mark item as read and `R`
mark item in all feeds as read.
The state of a toggle could be shown in the bottom line, which shows the
current binds, on the right side:

`q:Quit ENTER:Open [all the other stuff] Mark read in all feeds: Off`

Or something similar.

**Comment by Minoru** [comment]
> would it be too much of a performance hit to search the DB for matching GUIDs where feedURLs share the same domain address?

I don't think I understand the idea... Items are considered to be the same if their GUIDs are the exact same. That's it. Why would we look at URLs of the feeds?

If you're thinking of a way to prevent two *different* feeds from clashing when their items *happen* to have the same GUIDs—well, what you describe might be a valid approach, and the check shouldn't be much of a performance hit (given that we'll have just a few items that have identical GUIDs, so not much cross-checking to do). But probability of two different feeds having items with identical GUIDs should be negligible (GUID means "**globally unique** identifier", after all).

I also don't see how this answers my question about displaying duplicate items in non-query feeds. Or you weren't trying to answer that?

Actually, I do have one idea regarding that: we can add a column displaying a number of copies each item has, much like `ls -l` does it for hard links. But that will only work in item view—we still need something for feed view, so that user understands what Newsbeuter will do to "Asian news" if he or she marks "Chinese news" as read. What do you guys think?

> that or separate commands

Separate commands sound like Newsbeuter's way, so I'll go with that; thanks! Maybe we'll rethink that when we gain tagging support—I can think of this thing as a prefix to ordinary functions like read, delete and flag (basically the same as tag prefix in Mutt).

> The state of a toggle could be shown in the bottom line

Both top and bottom lines seem so packed with stuff that I'm not sure users will notice if something changes there. Furthermore, bottom line is supposed to be static, I guess, because it contains hints; Midnight Commander doesn't change its bottom line, does it?

I really like how Vim behaves when pastetoggle is pressed, but we don't have a separate line for current mode (nor do we have modes :), so that option's out. If we go with a column for number of links (see above), we could hide or display the column—that's readily noticeable because item titles will shift a bit. Any suggestions?

**Comment by kriskropd** [comment]
> I don't think I understand the idea... Items are considered to be the same if their GUIDs are the exact same. That's it. Why would we look at URLs of the feeds?

I thought it might decrease the performance hit if you only searched similar feed URLs rather than globally searching the entire db for duplicate GUIDs. I'm not really familiar with sqlite optimization, but I know that *usually* global DB searching can be ugly, performance-wise, if the DB is big enough. In my mind, I saw the duplication search to be an on-the-spot solution that would need to be fast.

The way this would assist in marking duplicates within the UI is: when the feeds update within newsbeuter, the UI could then mark which GUIDs were found multiple times within the DB. It would only be temporarily stored information, similar to the "illusion" that is unread duplicate items in current version. (Referencing yesterday's example where upon refresh an rss feed jumped from 3/3 to 10/10 after refreshing and including duplicates that were excluded previously - ignoring cache.db.)

-------
I like your idea of borrowing the hard links solution fro `ls`. Maybe the feed list should vaguely indicate which feeds contain duplicates using a letter notation like 'D' and set it right beside the letter notation 'N' in the second column which indicates a feed contains unread items. I would think trying to indicate that Chinese news overlaps with Asian news within the feed list itself would only promote clutter that the user probably won't care about. When the user is actually looking at an article, the duplication information could be stored in the head information area.

It also may be better to litely denote duplication on the item list too, actually - similar to the "unread" 'N' notation column. For example, if a user were looking at a query-feed - the duplicates would be stacked one-atop-the other already. To have them expanded within the item_list again would just be noise. And even in normal item list view, knowing that one article in Chinese feed re-occurs in the Asian news isn't really important beyond simply knowing that the article is a duplicate. Unless there is a scenario I'm not considering. I'm struggling to find scenarios where a user actually wants to treat duplicates individually. Thus, I keep thinking that minimalism is the way to handle duplicatation-indication until actually viewing a duplicated article in the article view.

**Comment by Minoru** [comment]
> I thought it might decrease the performance hit if you only searched similar feed URLs rather than globally searching the entire db for duplicate GUIDs.

Oh, that's right. Well, feedurl is no longer stored in the same table as items (and their GUIDs) are, so it becomes complicated: it is indeed faster to join tables and then match on both GUID and feedurl, but only when we got a huge number of items in DB. Otherwise, it's faster to just go look by GUID. Thanks for pointing this out; I'll make sure to pay attention to this as I write queries for new DB schema.

Each duplicate will be just a pair of numbers—feed id and item id—so no need for "illusions": it's quite lightweight already.

> I would think trying to indicate that Chinese news overlaps with Asian news within the feed list itself would only promote clutter that the user probably won't care about.

Seems like it. Good point!

> When the user is actually looking at an article, the duplication information could be stored in the head information area.

Oh, I didn't think of that! Yes, something like "Also appears in: <list of feed titles>" will prove quite useful. Users will even be able to figure out "what these numbers next to item titles are" just by looking how many feeds are listed when they open the item.

> It also may be better to litely denote duplication on the item list too, actually - similar to the "unread" 'N' notation column. For example, if a user were looking at a query-feed - the duplicates would be stacked one-atop-the other already. To have them expanded within the item_list again would just be noise.

Didn't get it, sorry. If we use `ls -l`-style numbers right before item titles, we shouldn't need any N-like flags, do we? And regarding query feeds: I think that with `ls -l`-like display we can just collapse all the duplicates into one entry. (That will mean we will have to decide what to put into "Feed" column, and the choice is bound to be arbitrary—the item appears in multiple feeds, but we can't decide which one of those is "primary"!)

Damn, just 23 hours since I opened the issue, and we already got the design almost fleshed out! Newsbeuter's community is made of pure awesomeness ☺

**Comment by kriskropd** [comment]
I actually edited my post before you quoted me, but I think it's easier to just admit I didn't consider collapsed information within only one, primary feed. Athough, I had considered it might just be easier to **not** select a primary feed, but instead just have them all indicated.

**Comment by Minoru** [comment]
When I talked about primary feed, it was in context of query feeds. They have this additional column, right before the items' titles, where the name of the feed from which the item came is displayed. That poses no problem if duplicates are listed multiple times in query feeds—they all will just have different feed titles, that's all. But if we try to collapse duplicates into one entry (e.g. one line), we need to decide which feed title to display in that additional column. We can't display them all lest we have no space left to display the items' title.

So either we don't collapse duplicates in query feeds at all, or we *arbitrarily* decide which feed will have its title displayed next to item's title.

**Comment by insteps** [comment]
Not sure if this would be easy and require lots of changes.
1. I was thinking of meta-status database/table (separate), that marks the individual items in db as read/deleted but under each RSS-Feeds category. By delete it means soft-delete.
2. On viewing each Feed the status' are fetched from this table.
3. There could be a global cleanup command for doing real-delete, that can also be run at command line. Kinda cleanup routines.  Issuing this command form CLI should reload/refresh the CLI Gui.

**Comment by insteps** [comment]
4. Maybe there could be also be duplicate item column in CLI/GUI ( easy to figure out from status table ), i.e item having more than 1 entry

**Comment by Minoru** [comment]
1. Yep, that's pretty much how I'm going to implement all this.
2. Yes, exactly.
3. What is the use-case for this?
4. Yep, we discussed this exact approach above, and I'm going to implement it as well.

**Comment by insteps** [comment]
Above (3) has some advantages.
1. Cleaning from CLI can be expensive and disrupt the flow of reader.
2. I can schedule it (cron), like,
a. Mark all read items having duplicate item as read.
b. Delete items older than N-days ( this is kinda a new feature )
3. More features - archive N-days, i.e move them to separate db, like db-<archivedOn>.sqlite

This will also help me in improving my frontend (Insteps/Nbreader)
I was thinking of adding multiple users who only have meta-data different, but same Data<items>.db
(global cleanup would be more like admin work, i.e to keep total disk quota... but those are just some mixed thoughts....)

**Comment by Minoru** [comment]
1. How so? And just to be clear, we're talking of removing all items that are marked as "soft-deleted" from the DB, right? That can be easily incorporated into cleanup routine that is run on quit (if `cleanup-on-quit` is enabled, and it is, by default).
2. a) Do you *really* need this? Maybe it will suffice to treat duplicate items as one, as proposed above? (When one of the duplicates is marked as read, all of them will be marked as read.) Read articles will be deleted on quit if `delete-read-articles-on-quit` is on (it's off by default). b) I thought we already had that! See `keep-articles-days` setting.
3. For stuff like that, I'd rather support exporting the contents of DB in some common format like JSON. Would that be okay, in your opinion?

I didn't know people are writing software on top of Newsbeuter. That's awesome! Feel free to open new issues if you need enhancements or features that can make your life easier. I don't promise to implement everything people request, but at least we can discuss things and maybe make both Newsbeuter and Nbreader better.

**Comment by insteps** [comment]
Yes, right I did miss out some features in .conf (keep-articles-days), nice !
I think currently solving duplicates with any of the 2 approaches, 
without increasing any overhead should be great.

Keeping the data and meta dbs separate would be nice and helpful.
Note, for meta simple key-value store can also be used.
Hope to see the next realease soonish.

Rest of features I would try add in 'Nbreader`.
Nbreader is not targeted for end user for now, but requires some admin setup. I wanted a centralized news/rss collection that can be read from any current browser, mobile or desktop. Seems it works as I thought. Next plan is to add flag/tag/bag(bookmark) to rss feeds.

**Comment by Dr-Terrible** [comment]
I prefer `yes` as the default. The lack of suck feature annoys me beyond measure :)

**Comment by Minoru** [comment]
Can't believe it's been a year since I opened this issue. I stopped working on this because we didn't have tests for our SQL, so I wasn't confident I'm not breaking anything. Now we have tests, so time to re-visit our year-old design.

By majority vote, `treat-identical-items-as-one` will be set to `yes` by default.

That's the only part of the original design that I still consider to be valid, though. The rest of it seems over-engineered to me. (Looks like @kriskropd was hinting at this all along, even mentioning minimalism at one point.)

**Keybind to toggle the setting**. I'm hesitant to add new command because then we'll have to keep it until the next major release. Defining two macros (one to set the option to `yes`, another to `no`) is clumsy but works just as well. This will be further simplified by a change to keybind and macro system that I didn't fully thought out yet, but do plan to implement.

**Displaying the state of the setting**. The need for this is obviated by either: 1) user having the setting set to `yes` or `no` at all times; 2) user having two macros to set the setting to desired state. (This might change in the future, but for now it's so.)

**Separate commands for, say, deleting *this* item and deleting *this and all its duplicates***. I know I said this sounds like Newsbeuter way, but this is just pure overhead that forces users to remember two sets of keys instead of one.

**Additional column à la `ls -l` (hidden when `yes`, shown when `no`)**. Geeky (yay, more data!) but a big chunk of work that I don't see much need for: this functionality never appeared in any real use-cases we discussed.

**"Also appears in: " line in the header of the article**. No supporting evidence that anyone needs this.

The only real case for `treat-identical-items-as-one no` was presented by @ntnn. It's covered by either:

* two macros as I described above. In this case @ntnn can unset the setting before marking his main feed as read.
* @ntnn slightly altering his reading order so that subfeeds are read before the main feed (thus by the time he marks the main feed read, there's no articles in subfeeds that he might want to keep unread.)

None of the alternative solutions we discussed made it any easier than this, but all of them required more code in Newsbeuter.

The only problem left open is query feeds: should we collapse duplicate items into one and if yes, how? I think I'll go with minimalism here as well: items won't be collapsed because it's easier (less code in Newsbeuter, no extra decisions required.)

If anyone sees any glaring omissions in this summary or decisions I made, do tell!

**Comment by tsipinakis** [comment]
>Keybind to toggle the setting. I'm hesitant to add new command because then we'll have to keep it until the next major release. Defining two macros (one to set the option to yes, another to no) is clumsy but works just as well.

This functionality already exists, for example I have this in my config to toggle mark-as-read-on-hover when <leader>m is pressed `macro m set mark-as-read-on-hover!`
